indra/media_plugins/cef/linux_volume_catcher_pw.cpp

```diff
@@ -0,0 +1,451 @@
+ * $/LicenseInfo$
+ * @endcond
+ */
+
+/*
+  The high-level design is as follows:
+  1) Connect to the PipeWire daemon
+  2) Find all existing and new audio nodes
+  3) Examine PID and parent PID's to see if it belongs to our process
+  4) If so, tell PipeWire to adjust the volume of that node
+  5) Keep a list of all audio nodes and adjust when we setVolume()
+ */
+
+#include "linden_common.h"
+
+#include "volume_catcher.h"
+
+#include <unordered_set>
+#include <mutex>
+
+extern "C" {
+#include <pipewire/pipewire.h>
+#include <spa/pod/builder.h>
+#include <spa/param/props.h>
+
+#include "apr_pools.h"
+#include "apr_dso.h"
+}
+
+#include "media_plugin_base.h"
+
+SymbolGrabber gSymbolGrabber;
+
+#include "linux_volume_catcher_pw_syms.inc"
+
+////////////////////////////////////////////////////
+
+class VolumeCatcherImpl
+{
+public:
+	VolumeCatcherImpl();
+	~VolumeCatcherImpl();
+
+	bool loadsyms(std::string pw_dso_name);
+	void init();
+	void cleanup();
+
+	void pwLock();
+	void pwUnlock();
+
+	void setVolume(F32 volume);
+	// void setPan(F32 pan);
+
+	void handleRegistryEventGlobal(
+		uint32_t id, uint32_t permissions, const char* type,
+		uint32_t version, const struct spa_dict* props
+	);
+
+	class ChildNode
+	{
+	public:
+		bool mActive = false;
+
+		pw_proxy* mProxy = nullptr;
+		spa_hook mNodeListener {};
+		spa_hook mProxyListener {};
+		VolumeCatcherImpl* mImpl = nullptr;
+
+		void updateVolume();
+		void destroy();
+	};
+
+	bool mGotSyms = false;
+
+	F32 mVolume = 1.0f; // max by default
+	// F32 mPan = 0.0f; // center
+
+	pw_thread_loop* mThreadLoop;
+	pw_context* mContext;
+	pw_core* mCore;
+	pw_registry* mRegistry;
+	spa_hook mRegistryListener;
+
+	std::unordered_set<ChildNode*> mChildNodes;
+	std::mutex mChildNodesMutex;
+};
+
+VolumeCatcherImpl::VolumeCatcherImpl()
+{
+	init();
+}
+
+VolumeCatcherImpl::~VolumeCatcherImpl()
+{
+	cleanup();
+}
+
+// static void debugClear()
+// {
+// 	auto file = fopen("/home/maki/git/firestorm-viewer/log.txt", "w");
+// 	fprintf(file, "\n");
+// 	fclose(file);
+// }
+
+// static void debugPrint(const char* format, ...)
+// {
+//     va_list args;
+//     va_start(args, format);
+// 	auto file = fopen("/home/maki/git/firestorm-viewer/log.txt", "a");
+//     vfprintf(file, format, args);
+// 	fclose(file);
+//     va_end(args);
+// }
+
+static void registryEventGlobal(
+	void *data, uint32_t id, uint32_t permissions, const char *type,
+	uint32_t version, const struct spa_dict *props
+)
+{
+	static_cast<VolumeCatcherImpl*>(data)->handleRegistryEventGlobal(
+		id, permissions, type, version, props
+	);
+}
+
+static const struct pw_registry_events REGISTRY_EVENTS = {
+	.version = PW_VERSION_REGISTRY_EVENTS,
+    .global = registryEventGlobal,
+};
+
+bool VolumeCatcherImpl::loadsyms(std::string pw_dso_name)
+{
+	return gSymbolGrabber.grabSymbols({ pw_dso_name });
+}
+
+void VolumeCatcherImpl::init()
+{
+	// debugClear();
+	// debugPrint("init\n");
```

Comment:
I am aware this can aid in future debugging. On the other hand I am not a fan of having a lot of commented out code commited (there is more of that later)

Reply:
Perhaps do something like this?
```
        std::string log_file = /home/maki/git/firestorm-viewer/log.txt
        LL_DEBUGS("VolumeCatched") << "Printing state to file: " << log_file;
	    debugClear(log_file );
	    debugPrint(log_file , "init\n");
        LL_CONT << LL_ENDL;
```